### PR TITLE
 PackageManagerDownloadWorker refactor: step 3

### PR DESCRIPTION
**Pull request by tiegz**
TL;DR: this fetches the latest version for Maven packages from maven_metadata.xml, instead of fetching all the versions and picking the newest one.

We need to use `latest_version()` for `mapping()`, so we need the latest version AFAIK.

In the next step, we can make `versions()` optional, so we can get project data without making N requests for the versions.

Here's the original commit that started fetching the latest version to use as the project's POM: https://github.com/librariesio/libraries.io/commit/3b77f008b27cd05870b0bb534c1236e0a1213e9d#diff-ac475c62f3e3faba6f62b8831d6d8b85af275fdb812328fd9cfaec0179718e20R39

### CODE REVIEW

File: app/models/package_manager/maven.rb [hunk 1]
```diff
@@ -230,20 +232,20 @@ def self.licenses(xml)
...
-        Project
-          .find_by(name: name, platform: formatted_name)
-          &.versions
-          &.max_by(&:published_at)
-          &.number
-      end
+    def self.latest_version(name)
+      maven_metadata = get_raw(MavenUrl.from_name(name, repository_base, NAME_DELIMITER).maven_metadata)
+      latest = Nokogiri::XML(maven_metadata)
+        .css("versioning > latest, versioning > release, metadata > version")
```
Comment (tiegz) [inline app/models/package_manager/maven.rb:238]: I checked all the Maven providers' examples and `versioning > latest` looks reliable. I did find one weird example that had 1 version but no `versioning > latest`, so I'm providing two fallbacks that could be there too.

File: app/models/package_manager/maven.rb [hunk 2]
```diff
@@ -230,20 +232,20 @@ def self.licenses(xml)
...
-        # TODO: this is in place to handle packages that are no longer on maven-repository.com
-        # this could be removed if we switched to a package data provider that supplied full information
-        Project
-          .find_by(name: name, platform: formatted_name)
-          &.versions
-          &.max_by(&:published_at)
-          &.number
-      end
+    def self.latest_version(name)
+      maven_metadata = get_raw(MavenUrl.from_name(name, repository_base, NAME_DELIMITER).maven_metadata)
```
Comment (tiegz) [inline app/models/package_manager/maven.rb:236]: This fetches the maven metadata separately from `version()` (so we'll call it twice -- should be cheap, but is a good use-case for maintaining state in PackageManager::Base, where we could cache the maven metadata in the instance).

File: app/models/package_manager/base.rb [hunk 3]
```diff
@@ -121,8 +121,6 @@ def self.update(name, sync_version: :all)
 
       save_dependencies(mapped_project) if self::HAS_DEPENDENCIES
       finalize_db_project(db_project)
-    rescue SystemExit, Interrupt
```
Comment (jsonperl) [inline app/models/package_manager/base.rb:124]: 👋

File: app/models/package_manager/maven.rb (hunk 2 above)
Comment (jsonperl) [inline app/models/package_manager/maven.rb:236]: seems useful to make a `.maven_metadata` method?

File: app/models/package_manager/maven.rb [hunk 4]
```diff
@@ -230,20 +232,20 @@ def self.licenses(xml)
...
+      latest = Nokogiri::XML(maven_metadata)
+        .css("versioning > latest, versioning > release, metadata > version")
+        .map(&:text)
+        .first
+
+      # TODO: this is in place to handle packages that are no longer on maven-repository.com
```
Comment (jsonperl) [inline app/models/package_manager/maven.rb:242]: We're not using maven-repository now right? Maybe just delete this, and if it fails we can explore. Also if we need it, that query below can be a single pluck instead of hydrating hundreds/thousands of version objects to grab a single version number.

Comment (jsonperl) [reply app/models/package_manager/maven.rb:242]: `Version.where(project: Project.where(name: project.name, platform: project.platform)).order(published_at: :desc).limit(1).pluck(:number)`

File: app/models/package_manager/maven.rb (hunk 2 above)
Comment (tiegz) [reply app/models/package_manager/maven.rb:236]: True!

**Review by jsonperl — APPROVED** [review]

**Review by mellisio — APPROVED** [review]
